docs/Stream-Protocol-Spec.md

```diff
@@ -0,0 +1,88 @@
+- client should be able to request arbitrary ranges from the server
+- client can be assumed to have some of the data already and therefore can opt in to selectivally request chunks based on their hashes
+
+As mentioned, the client is typically expected to have some of the data in the stream. to mitigate duplicate data transmission the stream protocol provides a configurable message roundtrip before batch delivery which allows the downstream peer to selectively request the chunks which it does not store at the time of the request.
+This comes, expectedly, at a certain price. Since delivery batches are pre-negotiated and do not rely on the mere benevolence of nodes, we can conclude that the delivery batches are optimsed for _urgency_ rather than for maximising batch utilisation (this is however, would be more apparent with unbounded streams).
+
+the protocol defines the notions of:
+- **stream** - data source which is composed of a sequence of hashes, referenced by monotonically increasing integers, with unguaranteed contiguity with respect to one particular stream.
+- **client** _(downstream peer)_ - the peer which is requesting data and does not posses it (client)
+- **server** _(upstream peer)_ - the peer that has the data and sends it to the downstream peer (server)
+- **range** - based on the notion of integer indexes we can define a range that designates an interval on the stream.
+- **batch** - a set of chunks constituting an interval in a range are called a batch, with a length not exceeding a ceiling value negotiated when establishing streams
+- **batch delivery** - end of batch delivery should be indicated by an explicit message from the server
+both offered and wanted go together - note this
+- **roundtrip** - a configurable extra message exchange (negotiated on initial message exchange) meant to mitigate and avoid requesting the same data from different peers. when the roundtrip is not used the stream is assumed to be continuous and the order of delivery should be guaranteed, or the stream should be particularly ordered. A roundtrip consists of:
+    - **offered hashes** - from the server to the client
+    - **wanted hashes** - at the discretion of the client in response to offered hashes
+
+responsibilities:
+- client is able to request a range but doesnt know how many results the interval will return from the server
+- client does not know if interval is continuous or has gaps
+- range is defined by client and should be strictly respected and followed by server
+- all intervals specified in protocol messages are closed (inclusive)
+- when roundtrip is configured - chunk deliveries can be handled concurrently (therefore their order is not guaranteed), but a server end-of-batch with topmost session index must be sent to signal the end of a batch
```

Comment:
how/where do you configure the roundtrip? i am looking at the msg specs and it is not clear, i don't see it.

Reply:
i actually had a hard time placing it, that's why its not there. thank you for touching this topic. the main guiding principle for me is to make `server`s lean as much as possible, and `configuring` a roundtrip means more state maintained on the server. i think that introducing a `GetRangeRoundtrip` message would be hence better